### PR TITLE
perf(pdf): :zap: vectorise the pdf from scipy norm

### PR DESCRIPTION
**Pull request by lilbobbytablez**
This speeds up the pdf function 300-1000x. A very small increase in memory usage for medium size data sets with default resolution.

This in turn speeds up cdf and predict. In my use case, I had 12 hidden layers, 8 guassians. `X.shape` was around 8000 and resolution was default. This went from 30 seconds to <0.1 seconds.

A small amount of cleaning including:
- commented out the second call to make mu, sigma and pi. I believe only one is needed.
- I also remove redundant imports.

### CODE REVIEW

**Comment by koaning** [comment]
Speedups are always nice, but I wonder if we might be able to add some sort of a test for this.  Maybe some sort of a before/after plot of a distribution maybe?

File: skmdn/__init__.py [hunk 1]
```diff
@@ -167,20 +166,24 @@ def pdf(self, X, resolution=100, y_min=None, y_max=None):
             ys: (resolution,)
         '''
         X = torch.FloatTensor(X)
-        with torch.no_grad():
-            pi, mu, sigma = self.model_(X)
+
+        # with torch.no_grad():
+        #     pi, mu, sigma = self.model_(X)
+
```
[the remark this answers is not in the document]
Comment (lilbobbytablez) [reply skmdn/__init__.py:172]: I wasn't sure which one was preferable. Any opinion on this?

**Comment by lilbobbytablez** [comment]
I've added a hacky notebook under `examples/mdn.ipynb`. I plan to remove this before merging. It shows a 140x speed up in this example (moons dataset with 10,000 samples).

The vectorised scipy norm function should be numerically identical to the slower looped version. The very small differences are likely due to random seeds. I played around with trying to fix these but with numpy, scipy and pytorch all being involved I quickly gave up on that.

I also cast `pi` to float64 so that it plays nicely with different versions of numpy. I realised my previous commit was written using numpy 2.0 so this broke when I reinstalled. Should be fixed for numpy<2 now.

**Comment by lilbobbytablez** [comment]
Hey! Is the added notebook clear enough? If not, let me know. It would be great to get this PR merged.

**Comment by koaning** [comment]
Ah sorry for the delay. Will have a look now.

**Review by koaning — APPROVED** [review]
Yeah this makes a tonne of sense. The `np.broadcast` utility should surely outperform those for-loops.

**Comment by koaning** [comment]
Thanks for the PR! I will make a few archival changes on my side and will push a new version after.

**Comment by lilbobbytablez** [comment]
No worries! I know you're a pretty busy guy 😄 and great. Looking forward to seeing it in!

File: skmdn/__init__.py [hunk 2]
```diff
@@ -167,20 +166,13 @@ def pdf(self, X, resolution=100, y_min=None, y_max=None):
...
-            y_min if y_min is not None else self.y_min_,
-            y_max if y_max is not None else self.y_max_,
-            resolution
-        )
-        pdf = np.zeros((pi.shape[0], resolution))
-        for i in range(pi.shape[0]):
-            for j in range(pi.shape[1]):
-                pdf[i] += norm(mu[i, j], sigma[i, j]).pdf(ys) * pi[i, j]
+
+        ys = np.linspace(y_min or self.y_min_, y_max or self.y_max_, resolution)
```
Comment (AhmedThahir) [inline skmdn/__init__.py:171]: Wouldn't line 171 will break if y_min=0 and we actually want y_min=0. I had previously made #2 to fix #1 😆

Kindly let me if I'm mistaken.

Comment (koaning) [reply skmdn/__init__.py:171]: Ah crud, you might be right. My bad! 

If you're up for it, feel free to make a PR. Maybe with a test?

Comment (AhmedThahir) [reply skmdn/__init__.py:171]: Any chance you can share reference on how to make a test? 

I'm unaware of how to do so.

**Pull request merged**